File: 20-actors/fuchi/methods/analyze.py
"""analyze.py — 扶持 (fuchi) end-to-end allocation membrane over the :representative seed.

Per ADR-2606052300 (R0 + R1 a/b/c). Runs each seed maintainer through the full pipeline:

    covenant check → sustenance envelope → tenure-weighted allocation (cash≡0)
                   → in-kind rail decomposition → governance gate
                   → [auto | sbt-vote (R1b: real 1 SBT=1 vote + 48h timelock) | council-lv7 | refused]
                   → R1a: provisioning intents to the real producing actors (dry-run, published=false)
                   → R1c: toritate ledgerEntry booking (cash≡0) + kanae-renderable flow graph

and emits an offline scorecard (Markdown) + derived datoms (edn). NO live disbursement /
provisioning / land grant / binding vote — all G10 (Council Lv6+ + operator). Dry-run only.

Stdlib only.
"""

from __future__ import annotations

import pathlib

import book
import couple as couple_mod
import live_gate
import provision as prov
import vote as vote_mod
from _edn import load_edn
from allocate import allocate, cohort_from_seed
from route import (
    gov_route,
    in_kind_coverage,
    rider_hit,
    route_envelope,
    touches_invariant,
)

_ROOT = pathlib.Path(__file__).resolve().parents[1]
_SEED = _ROOT / "data" / "seed-sustenance-graph.kotoba.edn"
_OUT = _ROOT / "methods" / "out"


def _kw(v) -> str:
    return str(v or "").lstrip(":").split("/")[-1].lower()


def _envelopes_for(seed: dict, did: str) -> list[dict]:
    return [e for e in seed.get(":envelope/batch", []) if e.get(":envelope/maintainer") == did]


def _ballots_for(seed: dict, did: str) -> dict:
    for v in seed.get(":gov/ballots", []):
        if v.get(":gov/maintainer") == did:
            return v
    return {}


def run(seed_path: pathlib.Path = _SEED) -> dict:
    seed = load_edn(seed_path)
    records = seed.get(":maintainer/batch", [])
    ceiling = int(seed.get(":graph/stage-ceiling-usd-micros-yr", 30_000_000_000))
    now_h = int(seed.get(":graph/now-hours", 0))

    cohort = cohort_from_seed(records)
    allocs = {a.maintainer_did: a for a in allocate(cohort, ceiling, instrument="sustenance")}
    note_of = {r.get(":maintainer/did"): r.get(":maintainer/note", "") for r in records}
    maintains_of = {r.get(":maintainer/did"): " ".join(r.get(":maintainer/maintains", []) or [])
                    for r in records}

    rows: list[dict] = []
    derived: list[dict] = []
    intents: list[prov.ProvisioningIntent] = []
    ledger: list[book.LedgerEntry] = []
    flows: list[book.FlowEdge] = []
    in_kind_by_actor: dict[str, int] = {}   # R1(d): accepted in-kind imputed per displacing actor
    as_of = 3000
    for r in records:
        as_of += 10
        did = r.get(":maintainer/did")
        cov = _kw(r.get(":maintainer/covenant", ":vowed"))
        env = _envelopes_for(seed, did)

        try:
            rails = route_envelope(env)
        except ValueError as exc:
            rows.append({"did": did, "covenant": cov, "route": ":refused-at-intake",
                         "imputed": "-", "in_kind": "-", "outcome": "refused", "note": str(exc)})
            continue

        imputed_total = sum(rl.imputed_usd_micros_yr for rl in rails)
        coverage = in_kind_coverage(rails)
        ctx = note_of.get(did, "") + " " + maintains_of.get(did, "")
        rider = rider_hit(ctx)
        inv = touches_invariant(ctx)
        route = gov_route(imputed_total, inv, rider)

        outcome = "pending"
        if route == "auto":
            outcome = "accepted"
        elif route == "sbt-vote":
            v = _ballots_for(seed, did)
            ballots = vote_mod.ballots_from_seed(v.get(":gov/votes", []))
            t = vote_mod.tally(ballots, int(v.get(":gov/opened-at-hours", 0)), now_h,
                               timelock_h=int(v.get(":gov/timelock-h", 48)))
            outcome = t["outcome"]
            route = f"sbt-vote {t['yes']}-{t['no']}/{t['timelock_h']}h{'✓' if t['finalizable'] else '…'}"
        elif route == "council-lv7":
            outcome = "pending"
        elif route == "refused":
            outcome = "refused"

        a = allocs.get(did)
        rows.append({
            "did": did, "covenant": cov, "route": ":" + route, "imputed": imputed_total,
            "in_kind": coverage, "share": getattr(a, "share", 0.0),
            "rank": getattr(a, "priority_rank", "-"), "floor": getattr(a, "floor_usd_micros_yr", 0),
            "outcome": outcome, "note": note_of.get(did, ""),
        })

        # R1 a/c — provision + book only for allocations that are accepted or council-pending
        if a is not None and outcome in ("accepted", "pending"):
            # R1(d) — accumulate this maintainer's in-kind imputed onto each actor they maintain
            in_kind_imputed = int(round(imputed_total * coverage))
            for act in (maintains_of.get(did, "").split() or []):
                in_kind_by_actor[act] = in_kind_by_actor.get(act, 0) + in_kind_imputed
            these_intents = prov.provision(rails, a.maintainer_did)
            these_ledger = book.book_toritate(rails, a.maintainer_did, a.maintainer_did)
            these_flows = book.flow_graph(rails, a.maintainer_did, a.maintainer_did)
            intents.extend(these_intents)
            ledger.extend(these_ledger)
            flows.extend(these_flows)
            derived.append({
                ":alloc/maintainer": did, ":alloc/instrument": ":" + a.instrument,
                ":alloc/share": a.share, ":alloc/priority-rank": a.priority_rank,
                ":alloc/floor-usd-micros-yr": a.floor_usd_micros_yr,
                ":alloc/cash-usd-micros": 0, ":alloc/server-held-key": False,
                ":gov/route": ":" + route.split()[0], ":gov/outcome": ":" + outcome,
                ":rail/coverage-in-kind": coverage,
                ":prov/intents": len(these_intents), ":book/entries": len(these_ledger),
                ":wb/as-of": as_of,
            })

    # R1(d) — Displacement-Dividend coupling: per-cohort earmark + G2 gate
    coupling: list[dict] = []
    for ev in couple_mod.events_from_seed(seed.get(":cohort/displacement", [])):
        em = couple_mod.earmark_from_surplus(ev)
        committed = in_kind_by_actor.get(ev.displacing_actor, 0)
        g = couple_mod.coupling_gate(ev, em, committed)
        coupling.append({"earmark": em, "gate": g})

    # R1(live) — show every outward leg's gate REFUSING by default (no operator flag / attestation /
    # Council / member signature in a dry run). This is the deliverable: the live path exists and is
    # refused unless fully gated. `env={}` ⇒ the operator process flag is absent.
    live_status = [live_gate.gate_status(live_gate.LiveGate(leg=leg), env={})
                   for leg in live_gate.LEG_POLICY]

    return {"rows": rows, "derived": derived, "intents": intents, "ledger": ledger,
            "flows": flows, "coupling": coupling, "live_status": live_status}


def _report(res: dict) -> str:
    out = [
        "# 扶持 (fuchi) — maintainer-sustenance allocation dry-run (R0 + R1 a/b/c)\n",
        "The charter-clean INVERSE of an investment fund over the `:representative` seed. "
        "No equity, no ROI, no exit; **cash≡0** everywhere. No live disbursement / provisioning / "
        "land grant / binding vote (G10).\n",
        "## Allocation routing + governance\n",
        "| maintainer | covenant | imputed USD/yr | in-kind | share | rank | floor USD/yr | route | outcome |",
        "|---|---|---|---|---|---|---|---|---|",
    ]
    for r in res["rows"]:
        imp = r["imputed"]
        imp_s = f"${imp/1_000_000:,.0f}" if isinstance(imp, int) else imp
        floor = r.get("floor", 0)
        floor_s = f"${floor/1_000_000:,.0f}" if isinstance(floor, int) else floor
        ik = r["in_kind"]
        ik_s = f"{ik*100:.0f}%" if isinstance(ik, float) else ik
        did = r["did"].split(":")[-1]
        out.append(f"| {did} | {r['covenant']} | {imp_s} | {ik_s} | {r.get('share', 0.0)} | "
                   f"{r.get('rank', '-')} | {floor_s} | {r['route']} | {r['outcome']} |")

    out.append("\n## R1(a) provisioning intents → real producing actors (dry-run, published=false)\n")
    out.append("| alloc | rail | provider | kind | imputed USD/yr | member-principal |")
    out.append("|---|---|---|---|---|---|")
    for i in res["intents"]:
        out.append(f"| {i.alloc_id.split(':')[-1]} | {i.rail_kind} | "
                   f"{i.provider_did.split(':')[-1]} | {i.provider_kind} | "
                   f"${i.imputed_usd_micros_yr/1_000_000:,.0f} | {'✓ (warifu 0%)' if i.member_principal else '—'} |")

    out.append("\n## R1(c) toritate ledgerEntry booking (cashStipendUsd ≡ 0)\n")
    out.append("| alloc | category | imputed USD/yr | cash |")
    out.append("|---|---|---|---|")
    for e in res["ledger"]:
        out.append(f"| {e.alloc_id.split(':')[-1]} | {e.category} | "
                   f"${e.imputed_usd_micros_yr/1_000_000:,.0f} | ${e.cash_usd_micros} |")

    out.append("\n## R1(c) kanae-renderable internal sustenance-flow graph\n")
    out.append("| from | to | class | imputed USD/yr | in-kind |")
    out.append("|---|---|---|---|---|")
    for f in res["flows"]:
        out.append(f"| {f.frm.split(':')[-1]} | {f.to.split(':')[-1]} | {f.flow_class} | "
                   f"${f.imputed_usd_micros_yr/1_000_000:,.0f} | {'✓' if f.in_kind else '— (member loan)'} |")

    out.append("\n## R1(d) Displacement-Dividend coupling (G2: no displacement without a funded cohort)\n")
    out.append("| cohort | displacing actor | gross USD/yr | tithe 10% | earmark 90% | funded | committed | admissible |")
    out.append("|---|---|---|---|---|---|---|---|")
    for c in res.get("coupling", []):
        em, g = c["earmark"], c["gate"]
        out.append(f"| {em.cohort_id} | {em.displacing_actor} | "
                   f"${em.gross_usd_micros_yr/1_000_000:,.0f} | ${em.tithe_usd_micros/1_000_000:,.0f} | "
                   f"${em.earmark_usd_micros_yr/1_000_000:,.0f} | {'✓' if em.funded else '—'} | "
                   f"${g['committed']/1_000_000:,.0f} | {'✓ admissible' if g['admissible'] else '✗ REFUSED'} |")

    out.append("\n## R1(live) outward-leg gate (default = REFUSED; G10)\n")
    out.append("Each live leg fires only when its operator flag + attestation + Council level + "
               "member signature are ALL present. In a dry run none are, so every leg is refused.\n")
    out.append("| leg | env flag | min Council | operator flag | attested | ratified | member-signed | admissible |")
    out.append("|---|---|---|---|---|---|---|---|")
    for s in res.get("live_status", []):
        c = s["conditions"]
        def _m(b):
            return "✓" if b else "✗"
        out.append(f"| {s['leg']} | `{s['env_flag']}` | Lv{s['min_council']} | "
                   f"{_m(c['operator_flag'])} | {_m(c['operator_attested'])} | "
                   f"{_m(c['council_ratified'])} | {_m(c['member_signed'])} | "
                   f"{'✓ admissible' if s['admissible'] else '✗ REFUSED'} |")

    out.append("\n## Invariants held\n")
    out.append("- **cash≡0** — every allocation / provisioning intent / ledgerEntry carries 0 cash (N1).")
    out.append("- **no investment vehicle** — instrument ∈ sustenance set; equity/debt/ROI/exit unrepresentable (G1).")
    out.append("- **1 SBT = 1 vote + 48h timelock** — escalated allocations finalize only after the window closes (R1b).")
    out.append("- **in-kind first** — provisioning routed to real producing actors; external fiat only as member-principal warifu 0% (G3).")
    out.append("- **booked, not paid** — toritate cashStipendUsd ≡ 0; no payroll/wage category (R1c).")
    out.append("- **coupled** — a displacement is admissible only against a funded cohort earmark; "
               "10% TitheRouter split exact (R1d).")
    out.append("- **outward-gated** — every live leg (provision/vote/book/couple) REFUSES unless "
               "operator flag + attestation + Council Lv6+/Lv7+ + member signature; default = refused (R1-live).\n")
    return "\n".join(out)


def main() -> int:
    res = run()
    _OUT.mkdir(parents=True, exist_ok=True)
    (_OUT / "allocation-dryrun.md").write_text(_report(res), encoding="utf-8")

    def emit(records: list[dict]) -> str:
        lines = []
        for d in records:
            kvs = []
            for k, v in d.items():
                if isinstance(v, bool):
                    kvs.append(f"{k} {str(v).lower()}")
                elif isinstance(v, str) and not v.startswith(":"):
                    kvs.append(f'{k} "{v}"')
                else:
                    kvs.append(f"{k} {v}")
            lines.append("  {" + " ".join(kvs) + "}")
        return "\n".join(lines)

    (_OUT / "allocations.kotoba.edn").write_text(
        "{:fuchi/derived\n [\n" + emit(res["derived"]) + "\n ]}\n", encoding="utf-8")
    (_OUT / "provisioning-intents.kotoba.edn").write_text(
        "{:fuchi/provisioning\n [\n" + emit([{
            ":prov/alloc": i.alloc_id, ":prov/rail-kind": ":" + i.rail_kind,
            ":prov/provider-did": i.provider_did, ":prov/provider-kind": ":" + i.provider_kind,
            ":prov/imputed-usd-micros-yr": i.imputed_usd_micros_yr,
            ":prov/member-principal": i.member_principal, ":prov/cash-usd-micros": 0,
            ":prov/server-held-key": False, ":prov/published": False,
        } for i in res["intents"]]) + "\n ]}\n", encoding="utf-8")
    (_OUT / "toritate-ledger.kotoba.edn").write_text(
        "{:fuchi/ledger\n [\n" + emit([{
            ":book/alloc": e.alloc_id, ":book/category": ":" + e.category,
            ":book/imputed-usd-micros-yr": e.imputed_usd_micros_yr,
            ":book/cash-usd-micros": 0, ":book/counterparty-did": e.counterparty_did,
        } for e in res["ledger"]]) + "\n ]}\n", encoding="utf-8")
    (_OUT / "kanae-flow.kotoba.edn").write_text(
        "{:fuchi/flow\n [\n" + emit([{
            ":flow/from": f.frm, ":flow/to": f.to, ":flow/class": ":" + f.flow_class,
            ":flow/imputed-usd-micros-yr": f.imputed_usd_micros_yr, ":flow/in-kind": f.in_kind,
        } for f in res["flows"]]) + "\n ]}\n", encoding="utf-8")
    (_OUT / "cohort-earmarks.kotoba.edn").write_text(
        "{:fuchi/coupling\n [\n" + emit([{
            ":earmark/cohort": c["earmark"].cohort_id,
            ":earmark/displacing-actor": c["earmark"].displacing_actor,
            ":earmark/gross-usd-micros-yr": c["earmark"].gross_usd_micros_yr,
            ":earmark/tithe-usd-micros": c["earmark"].tithe_usd_micros,
            ":earmark/earmark-usd-micros-yr": c["earmark"].earmark_usd_micros_yr,
            ":earmark/funded": c["earmark"].funded,
            ":couple/committed-usd-micros-yr": c["gate"]["committed"],
            ":couple/admissible": c["gate"]["admissible"],
        } for c in res["coupling"]]) + "\n ]}\n", encoding="utf-8")
    (_OUT / "live-gate-status.kotoba.edn").write_text(
        "{:fuchi/live-gate\n [\n" + emit([{
            ":gate/leg": ":" + s["leg"], ":gate/env-flag": s["env_flag"],
            ":gate/min-council": s["min_council"],
            ":gate/operator-flag": s["conditions"]["operator_flag"],
            ":gate/operator-attested": s["conditions"]["operator_attested"],
            ":gate/council-ratified": s["conditions"]["council_ratified"],
            ":gate/member-signed": s["conditions"]["member_signed"],
            ":gate/admissible": s["admissible"],
        } for s in res.get("live_status", [])]) + "\n ]}\n", encoding="utf-8")

    print(_report(res))
    print(f"\nwrote allocation-dryrun.md + allocations / provisioning-intents / "
          f"toritate-ledger / kanae-flow / cohort-earmarks / live-gate-status .kotoba.edn → {_OUT}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: 20-actors/fuchi/methods/analyze.clj
;; analyze.clj — 扶持 (fuchi) end-to-end allocation membrane over the :representative seed.
;;
;; Clojure port of analyze.py `run` (ADR-2606052300 R0 + R1 a/b/c/d), Wave 1 of the clj-native
;; migration (ADR-2606142300) — the FINAL fuchi method, completing the actor + Wave 1. Runs each
;; seed maintainer through the full pipeline over the now-clj methods:
;;
;;   covenant check → sustenance envelope → tenure-weighted allocation (cash≡0)
;;     → in-kind rail decomposition → governance gate
;;     → [auto | sbt-vote (real 1 SBT=1 vote + 48h timelock) | council-lv7 | refused]
;;     → provisioning intents to real producing actors (dry-run, published=false)
;;     → toritate ledgerEntry booking (cash≡0) + kanae-renderable flow graph
;;     → Displacement-Dividend cohort coupling (G2: no displacement without a funded cohort)
;;
;; NO live disbursement / provisioning / land grant / binding vote — all dry-run (G10). NOTE: only
;; the pipeline `run` is ported; analyze.py's `_report`/`main` are skipped (they read live-gate
;; condition keys the R2 autonomous gate no longer emits — i.e. that Markdown renderer is dead code
;; against the current live_gate). A clean `scorecard` summary is provided instead. stdlib only.
(ns fuchi.methods.analyze
  (:require [clojure.edn :as edn]
            [clojure.string :as str]
            [clojure.java.io :as io]
            [fuchi.methods.allocate :as al]
            [fuchi.methods.route :as rt]
            [fuchi.methods.vote :as vote]
            [fuchi.methods.book :as book]
            [fuchi.methods.provision :as prov]
            [fuchi.methods.couple :as couple]
            [fuchi.methods.live-gate :as lg])
  (:import [java.math BigDecimal RoundingMode]))

(def default-seed "20-actors/fuchi/data/seed-sustenance-graph.kotoba.edn")

(defn- kw [v] (-> (str (or v "")) (str/replace #"^:+" "") (str/split #"/") last str/lower-case))
(defn- round-int [x] (.longValue (.setScale (BigDecimal/valueOf (double x)) 0 RoundingMode/HALF_EVEN)))

(defn- envelopes-for [seed did]
  (filterv #(= (:envelope/maintainer %) did) (:envelope/batch seed)))

(defn- ballots-for [seed did]
  (or (first (filter #(= (:gov/maintainer %) did) (:gov/ballots seed))) {}))

(defn run
  "Run the full allocation pipeline over the seed. Returns
   {:rows :derived :intents :ledger :flows :coupling :live-status}."
  ([] (run default-seed))
  ([seed-path]
   (let [seed       (edn/read-string (slurp (io/file seed-path)))
         records    (:maintainer/batch seed)
         ceiling    (long (:graph/stage-ceiling-usd-micros-yr seed 30000000000))
         now-h      (long (:graph/now-hours seed 0))
         cohort     (al/cohort-from-seed records)
         allocs     (into {} (map (juxt :maintainer-did identity) (al/allocate cohort ceiling 0 "sustenance")))
         note-of    (into {} (map (fn [r] [(:maintainer/did r) (get r :maintainer/note "")]) records))
         maintains-of (into {} (map (fn [r] [(:maintainer/did r)
                                             (str/join " " (map str (get r :maintainer/maintains [])))]) records))]
     (loop [rs records, as-of 3000
            acc {:rows [] :derived [] :intents [] :ledger [] :flows [] :in-kind-by-actor {}}]
       (if (empty? rs)
         (let [coupling (mapv (fn [ev]
                                (let [em (couple/earmark-from-surplus ev)
                                      committed (get (:in-kind-by-actor acc) (:displacing-actor ev) 0)]
                                  {:earmark em :gate (couple/coupling-gate ev em committed)}))
                              (couple/events-from-seed (:cohort/displacement seed)))
               live-status (mapv #(lg/gate-status (lg/make-gate {:leg %})) (keys lg/leg-policy))]
           (-> acc (dissoc :in-kind-by-actor) (assoc :coupling coupling :live-status live-status)))
         (let [r       (first rs)
               as-of   (+ as-of 10)
               did     (:maintainer/did r)
               cov     (kw (get r :maintainer/covenant ":vowed"))
               env     (envelopes-for seed did)]
           (if-let [rails (try (rt/route-envelope env) (catch Exception e {::refused (.getMessage e)}))]
             (if (::refused rails)
               (recur (rest rs) as-of
                      (update acc :rows conj {:did did :covenant cov :route ":refused-at-intake"
                                              :imputed "-" :in-kind "-" :outcome "refused" :note (::refused rails)}))
               (let [imputed-total (reduce + 0 (map :imputed-usd-micros-yr rails))
                     coverage (rt/in-kind-coverage rails)
                     ctx      (str (note-of did) " " (maintains-of did))
                     rider    (rt/rider-hit ctx)
                     inv      (rt/touches-invariant ctx)
                     route0   (rt/gov-route imputed-total inv rider)
                     [outcome route]
                     (case route0
                       "auto"        ["accepted" route0]
                       "council-lv7" ["pending" route0]
                       "refused"     ["refused" route0]
                       "sbt-vote"    (let [v (ballots-for seed did)
                                           ballots (vote/ballots-from-seed (get v :gov/votes []))
                                           t (vote/tally ballots (long (get v :gov/opened-at-hours 0)) now-h
                                                         (long (get v :gov/timelock-h 48)) vote/default-quorum)]
                                       [(:outcome t)
                                        (str "sbt-vote " (:yes t) "-" (:no t) "/" (:timelock-h t) "h"
                                             (if (:finalizable t) "✓" "…"))])
                       ["pending" route0])
                     a (allocs did)
                     row {:did did :covenant cov :route (str ":" route) :imputed imputed-total
                          :in-kind coverage :share (:share a 0.0) :rank (:priority-rank a "-")
                          :floor (:floor-usd-micros-yr a 0) :outcome outcome :note (note-of did)}]
                 (if (and a (#{"accepted" "pending"} outcome))
                   (let [in-kind-imputed (round-int (* imputed-total coverage))
                         acts (remove str/blank? (str/split (maintains-of did) #"\s+"))
                         these-intents (prov/provision rails (:maintainer-did a) #{})
                         these-ledger  (book/book-toritate rails (:maintainer-did a) (:maintainer-did a))
                         these-flows   (book/flow-graph rails (:maintainer-did a) (:maintainer-did a))]
                     (recur (rest rs) as-of
                            (-> acc
                                (update :rows conj row)
                                (update :in-kind-by-actor #(reduce (fn [m act] (update m (symbol act) (fnil + 0) in-kind-imputed)) % acts))
                                (update :intents into these-intents)
                                (update :ledger into these-ledger)
                                (update :flows into these-flows)
                                (update :derived conj
                                        {:alloc/maintainer did :alloc/instrument (str ":" (:instrument a))
                                         :alloc/share (:share a) :alloc/priority-rank (:priority-rank a)
                                         :alloc/floor-usd-micros-yr (:floor-usd-micros-yr a)
                                         :alloc/cash-usd-micros 0 :alloc/server-held-key false
                                         :gov/route (str ":" (first (str/split route #"\s+")))
                                         :gov/outcome (str ":" outcome) :rail/coverage-in-kind coverage
                                         :prov/intents (count these-intents) :book/entries (count these-ledger)
                                         :wb/as-of as-of}))))
                   (recur (rest rs) as-of (update acc :rows conj row)))))
             (recur (rest rs) as-of acc))))))))

(defn scorecard
  "A clean dry-run summary (replaces analyze.py's R2-broken Markdown _report)."
  [res]
  (str/join "\n"
    (concat
     ["# 扶持 (fuchi) — maintainer-sustenance allocation dry-run (R0 + R1 a/b/c/d)"
      ""
      (str "- maintainers: " (count (:rows res))
           " · accepted: " (count (filter #(= "accepted" (:outcome %)) (:rows res)))
           " · pending: " (count (filter #(= "pending" (:outcome %)) (:rows res)))
           " · refused: " (count (filter #(= "refused" (:outcome %)) (:rows res))))
      (str "- provisioning intents: " (count (:intents res))
           " · ledger entries: " (count (:ledger res))
           " · flow edges: " (count (:flows res)))
      (str "- coupling: " (count (filter #(get-in % [:gate :admissible]) (:coupling res)))
           "/" (count (:coupling res)) " cohorts admissible")
      "" "| maintainer | covenant | route | outcome |" "|---|---|---|---|"]
     (map (fn [r] (str "| " (last (str/split (str (:did r)) #":")) " | " (:covenant r)
                       " | " (:route r) " | " (:outcome r) " |")) (:rows res)))))
